feat: Setup Kotest runner with output

NickHackman opened the pull request:
:warning: This is a work in progress :warning: 

Relates to #65 

## Goals

- Have an executable / gradle plugin
- Execute Kotest tests
- Standardize output for neotest-kotlin usage
- integrate with neotest-kotlin
- remove Lua output parsing

NickHackman committed:
feat: gradle init neotest-kotlin

Unsure on the name here, but just using this as of right now and we can
change it later

NickHackman committed:
build(deps): remove guava

NickHackman committed:
rebase onto setup

NickHackman committed:
feat: basic Kotest runner

Simple kotest test runner that uses the TestEngineListener to listen to
all tests output in order to construct a JSON report. Currently, the
specs that it's finding are hardcoded.

NickHackman committed:
build: show events and stdout

NickHackman committed:
build: remove JUnit

NickHackman committed:
build: add kotlin reflect

NickHackman committed:
build: add all Kotest dependencies

NickHackman committed:
refactor: TestNode represents a Container and Test

NickHackman committed:
feat: Kotest test running

This is missing some features and lacks error handling when Kotest fails
to run the tests in the first place.

NickHackman committed:
test: TestNodeSpec

NickHackman committed:
test: KotestTestRunner functional test

NickHackman committed:
test: KotestTestRunner unit tests

NickHackman commented on neotest-kotlin/app/src/main/kotlin/io/github/codymikol/neotestkotlin/App.kt:
```diff
@@ -0,0 +1,21 @@
+    val classes: List<KClass<*>> =
+        listOf(
+            Class.forName("org.example.KotestFunSpec").kotlin,
+        )
```
TODO here, need to actually allow for passing in the classes to run

NickHackman commented on neotest-kotlin/app/src/main/kotlin/io/github/codymikol/neotestkotlin/framework/TestFrameworkRunner.kt:
```diff
@@ -0,0 +1,52 @@
+                val result = runner.run(runnableClasses)
+                when (result) {
+                    is TestRunResult.Success -> result.report
+                    is TestRunResult.Failure -> TODO()
```
No idea what we should do when this fails, I imagine we want to spit out some JSON that explains what went wrong here?

codymikol replied:
I think that makes sense, on the lua side we can probably handle that by printing that into the overall output and failing all test cases with a generic error that basically says to look at the output for more information.

It doesn't look like you'd normally get this output and that it would either be a mistake on our end, or an internal kotest error.

NickHackman commented on neotest-kotlin/app/src/main/kotlin/io/github/codymikol/neotestkotlin/framework/TestNode.kt:
```diff
@@ -0,0 +1,182 @@
+ * other namespaces or tests.
+ */
+@OptIn(ExperimentalSerializationApi::class) // JsonClassDiscriminator is an experimental annotation for now
+@JsonClassDiscriminator("type")
+@Serializable
```
I'm using Kotlinx serialization here, which is nice and all, but have some caveats from good ole Jackson. Namely, the fields in `TestNode.Container` for `status` and `duration` aren't evaluated because they're not backed fields. I'll probably pivot away from this if there's no way to handle this.

[docs](https://github.com/Kotlin/kotlinx.serialization/blob/master/docs/basic-serialization.md#backing-fields-are-serialized)

NickHackman commented on neotest-kotlin/app/src/main/kotlin/io/github/codymikol/neotestkotlin/framework/TestNode.kt:
```diff
@@ -0,0 +1,182 @@
+                            error?.let {
+                                val traceOrigin = it.stackTrace?.firstOrNull()
+
+                                Failure.Error(
+                                    message = it.message,
+                                    lineNumber = traceOrigin?.lineNumber,
+                                    filename = traceOrigin?.fileName,
+                                )
```
This mostly works, outside of soft assertions which seem to reference a lineNumber that is higher than the file. Maybe just need to specially handle that?

NickHackman commented on neotest-kotlin/app/src/test/kotlin/io/github/codymikol/neotestkotlin/framework/kotest/KotestExampleSpec.kt:
```diff
@@ -0,0 +1,70 @@
+import io.kotest.matchers.ints.shouldBeOdd
+import io.kotest.matchers.shouldBe
+
+open class KotestExampleSpec :
```
I have no idea on how to make this Spec runnable via the functional test, but not run by Kotest itself. I'll probably need to dig through their code base to see how they do this sort of thing

NickHackman replied:
Oh yeah, just ignore the test. Done in

https://github.com/codymikol/neotest-kotlin/commit/424fc5a1d9e66604743701ac67827b319a26a8e8

NickHackman commented on neotest-kotlin/app/src/test/kotlin/io/github/codymikol/neotestkotlin/framework/kotest/KotestTestRunnerFunctionalSpec.kt:
```diff
@@ -0,0 +1,164 @@
+                    [
+                      {
+                        "name": "io.github.codymikol.neotestkotlin.framework.kotest.KotestExampleSpec",
+                        "nodes": [
```
Don't love the "nodes" naming here, but we effectively have a tree. I also don't know a good word for either a Container or a Test

NickHackman commented on neotest-kotlin/app/src/test/kotlin/io/github/codymikol/neotestkotlin/framework/kotest/KotestTestRunnerFunctionalSpec.kt:
```diff
@@ -0,0 +1,164 @@
+                            "type": "test",
+                            "name": "fail",
+                            "status": {
+                              "status": "failure",
```
The whole `stackTrace` is included here as well, but I omitted it for testing purposes. Going to add validation to ensure that the key exists at least

NickHackman replied:
https://github.com/codymikol/neotest-kotlin/pull/70/commits/45988d03d39b5a648055b5919752a7c4c427bc96

NickHackman committed:
test: assert duration and stackTrace exist on tests

NickHackman committed:
refactor: use AbstractTestEngineListener

NickHackman committed:
test: don't run Example specs

NickHackman committed:
refactor: neotest-kotlin -> kotlin-test-launcher

Better name for the project

NickHackman committed:
refactor: extract out core subproject

This will be the base for the gradle plugin and keep the core logic of
running and discovering based on Frameworks opaque and following a
contract.

NickHackman committed:
refactor: set up basic gradle plugin

NickHackman committed:
refactor(core): runBlocking rather than suspend

This makes it easier to use in the context of the gradle plugin

Commit:
feat: add the maven scaffolding

this is just a simple maven plugin that can eventually
hook into the core lib

Fixes N/A

Commit:
chore: moved maven -> maven-plugin

for consistency

Fixes N/A

NickHackman committed:
build: publish to mavenLocal + jackson

NickHackman committed:
refactor: don't use kotlinx serialization instead use Jackson

NickHackman committed:
feat: gradle-plugin

Takes in classes via -Pclasses=org.example.KotestFunSpec and writes to
the output file in the build

NickHackman commented on kotlin-test-launcher/gradle-plugin/src/main/kotlin/io/github/codymikol/kotlintestlauncher/plugin/KotlinTestLauncherPlugin.kt:
```diff
@@ -0,0 +1,23 @@
+
+            classes.set(project.properties["classes"]?.toString())
+
+            outputFile.convention(project.layout.buildDirectory.file("$name/output-${UUID.randomUUID()}.json"))
```
Currently just dumping to a randomly named output JSON file, but we should be good to provide a custom file via `-PoutputFile=/path/to/output/file/here.json`

NickHackman committed:
fix(test-logging.init.gradle.kts): remove test changes for custom plugin

NickHackman committed:
feat(gradle-plugin): provide custom outputFile to gradle-plugin

NickHackman committed:
refactor: local functions

NickHackman committed:
chore(gitignore): .idea

NickHackman committed:
feat: log out outputFile

NickHackman committed:
feat: use kotlin-test-launcher plugin output instead of gradle

NickHackman committed:
test: TestNode tests

NickHackman committed:
test: update

NickHackman committed:
refactor: remove output_parser entirely

This now leverages the kotlin-test-launcher plugin

NickHackman committed:
test: output/init.lua

NickHackman committed:
fix: remove output_parser import

NickHackman committed:
refactor: TestNode.status.status -> TestNode.status.type

Feels more clear than having two nested statuses

NickHackman committed:
fix: use Jackson over kotlinx serialization

NickHackman committed:
ci: kotlin-test and lua-test workflows

NickHackman committed:
fix: exclude abstract Kotest specs

NickHackman committed:
fix: never allow plugin to be up to date

NickHackman committed:
build: use same version of kotlin

NickHackman committed:
build(deps): add asm

NickHackman committed:
feat: support executing entire packages

NickHackman committed:
style: ktlint format

NickHackman committed:
refactor: use Set rather than List

Prevent execution of multiple of the same class

NickHackman committed:
feat: kotlin-test-launcher -> kotlin-test and run from root project

NickHackman commented:
v2.0.0 is in the works, this is no longer necessary.
